crowbar_framework/lib/crowbar/backup/export.rb

```diff
@@ -0,0 +1,180 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef"
+require "yaml_db"
+
+module Crowbar
+  class Backup
+    class Export < Base
+      attr_accessor :path
+
+      def initialize(path)
+        self.path = path
+      end
+
+      def export
+        clients
+        nodes
+        roles
+        databags
+        db
+        crowbar
+      end
+
+      def clients
+        chef(
+          "client",
+          "::Chef::ApiClient"
+        )
+      end
+
+      def nodes
+        chef(
+          "node",
+          "::Chef::Node"
+        )
+      end
+
+      def roles
+        chef(
+          "role",
+          "::Chef::Role"
+        )
+      end
+
+      def databags
+        logger.debug "Backing up databags"
+
+        data_dir = workdir.join("knife", "databags")
+        data_dir.mkpath unless data_dir.directory?
+
+        Chef::DataBag.list.each do |name, url|
+          bag_dir = data_dir.join(name)
+          bag_dir.mkpath unless bag_dir.directory?
+
+          Chef::DataBag.load(name).each do |item, item_url|
+            logger.debug "Backing up databag #{name}/#{item}"
+
+            record = Chef::DataBagItem.load(
+              name,
+              item
+            )
+
+            bag_dir.join("#{item}.json").open("w") do |file|
+              file.write JSON.pretty_generate(
+                record
+              )
+            end
+          end
+        end
+      end
+
+      def db
+        logger.debug "Backing up database"
+
+        data_dir = workdir.join("crowbar")
+        data_dir.mkpath unless data_dir.directory?
+
+        SerializationHelper::Base.new(
+          YamlDb::Helper
+        ).dump(
+          data_dir.join(
+            "database.#{YamlDb::Helper.extension}"
+          )
+        )
+      end
+
+      def crowbar
+        logger.debug "Backing up Crowbar Files"
+
+        data_dir = workdir.join("crowbar")
+        ["tftp", "etc", "crowbar", "root"].each do |folder|
+          data_dir.join(folder).mkpath
+        end
+
+        Crowbar::Backup::Base.export_files.each do |filemap|
+          if filemap[0] =~ /resolv.conf/
+            data_dir.join("#{filemap[1]}").open("w") do |file|
+              forwarders.each do |fw|
+                file.write("nameserver #{fw}")
+              end
+            end
+          else
+            FileUtils.cp_r(filemap[0], data_dir.join(filemap[1]))
+          end
+        end
+
+        data_dir.join("version").open("w") do |file|
+          file.write(ENV["CROWBAR_VERSION"])
+        end
+      end
+
+      protected
+
+      def forwarders
```

Comment:
Just a small note (I understand we'll tackle that in a followup PR): in the old backup tool, we were actually checking for the existence of the file. So that was slightly better.